File: extensions/scripts/deconstruct_topics.py
#!/usr/bin/env python3
"""
deconstruct_topics_fire.py

Deconstruct Obsidian topic notes into append-safe finding notes.

Principles:
- Obsidian frontmatter is the database.
- Pandoc parses source frontmatter and Markdown into an AST.
- H1 blocks are the atomic split boundary.
- Pandoc applies the finding-note template.
- Pipeline writeback appends drafts; it never overwrites original notes.

Requires:
    pip install fire pyyaml
    pandoc on PATH

Typical use:
    python deconstruct_topics_fire.py deconstruct Research/ ImportedFindings/

Dry run:
    python deconstruct_topics_fire.py deconstruct Research/ ImportedFindings/ --dry_run=True

With a custom template:
    python deconstruct_topics_fire.py deconstruct Research/ ImportedFindings/ --template=fnd_template.pandoc.md
"""

from __future__ import annotations

import datetime as dt
import hashlib
import json
import re
import shutil
import subprocess
import sys
import tempfile
import unicodedata
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Iterable

import yaml

try:
    import fire
except ModuleNotFoundError as exc:  # fail loud, but with the one useful instruction
    raise SystemExit("fatal: Fire is not installed. Install with: pip install fire") from exc


FINDING_PREFIX = "fnd__"
TOPIC_PREFIX = "tpc__"
DEFAULT_FROM_FORMAT = (
    "markdown"
    "+yaml_metadata_block"
    "+wikilinks_title_after_pipe"
    "+bracketed_spans"
    "+fenced_divs"
)
DEFAULT_TO_FORMAT = DEFAULT_FROM_FORMAT

DEFAULT_TEMPLATE = r'''---
id: "$id$"
kind: finding
topic: "$topic$"
status: "$status$"
verification_status: "$verification_status$"
source_status: "$source_status$"
source_topic_file: "$source_topic_file$"
source_heading: "$source_heading$"
source_index: $source_index$
input_sha256: "$input_sha256$"
last_pipeline_run: "$last_pipeline_run$"
updated: "$updated$"
tags:
$for(tags)$
  - "$tags$"
$endfor$
---

# $title$

## Original note

$body$

## Pipeline drafts

<!-- Append generated drafts here. Do not overwrite Original note. -->

## Accepted version

<!-- Human-owned final/current version for topic transclusion. -->
'''


@dataclass(frozen=True)
class ParsedDocument:
    api_version: list[int]
    meta: dict[str, Any]
    blocks: list[dict[str, Any]]


@dataclass(frozen=True)
class Section:
    index: int
    title: str
    blocks: list[dict[str, Any]]


@dataclass(frozen=True)
class RenderedFinding:
    path: Path
    source: Path
    title: str
    note: str
    status: str


def fail(message: str) -> None:
    raise SystemExit(f"fatal: {message}")


def require_pandoc() -> None:
    if shutil.which("pandoc") is None:
        fail("pandoc is not available on PATH")


def run_pandoc(args: list[str], *, stdin: str) -> str:
    proc = subprocess.run(
        ["pandoc", *args],
        input=stdin,
        text=True,
        capture_output=True,
        check=False,
    )
    if proc.returncode != 0:
        fail("pandoc failed:\n" + proc.stderr.strip())
    return proc.stdout


def pandoc_json(markdown: str, *, from_format: str) -> dict[str, Any]:
    return json.loads(run_pandoc(["--from", from_format, "--to", "json"], stdin=markdown))


def parse_markdown(path: Path, *, from_format: str) -> ParsedDocument:
    raw = path.read_text(encoding="utf-8")
    ast = pandoc_json(raw, from_format=from_format)
    return ParsedDocument(
        api_version=ast.get("pandoc-api-version", [1, 23]),
        meta=ast.get("meta", {}),
        blocks=ast.get("blocks", []),
    )


def ast_to_markdown(
    *,
    blocks: list[dict[str, Any]],
    api_version: list[int],
    to_format: str,
    wrap: str,
) -> str:
    doc = {
        "pandoc-api-version": api_version,
        "meta": {},
        "blocks": blocks,
    }
    return run_pandoc(
        ["--from", "json", "--to", to_format, "--wrap", wrap],
        stdin=json.dumps(doc),
    ).strip()


def render_template(
    *,
    body_markdown: str,
    metadata: dict[str, Any],
    template_path: Path,
    from_format: str,
    to_format: str,
    wrap: str,
) -> str:
    with tempfile.TemporaryDirectory() as td:
        meta_path = Path(td) / "metadata.yml"
        meta_path.write_text(
            yaml.safe_dump(metadata, sort_keys=False, allow_unicode=True),
            encoding="utf-8",
        )
        rendered = run_pandoc(
            [
                "--from", from_format,
                "--to", to_format,
                "--standalone",
                "--template", str(template_path),
                "--metadata-file", str(meta_path),
                "--wrap", wrap,
            ],
            stdin=body_markdown,
        )
    return rendered.strip() + "\n"


def meta_scalar(meta_value: Any) -> str:
    """Small Pandoc Meta -> string converter for source frontmatter values."""
    if not isinstance(meta_value, dict):
        return ""
    tag = meta_value.get("t")
    value = meta_value.get("c")
    if tag == "MetaString":
        return str(value)
    if tag == "MetaBool":
        return "true" if value else "false"
    if tag == "MetaInlines":
        return stringify_inlines(value)
    if tag == "MetaBlocks":
        return stringify_blocks(value)
    if tag == "MetaList":
        return ", ".join(meta_scalar(item) for item in value)
    return ""


def source_frontmatter(parsed: ParsedDocument) -> dict[str, str]:
    return {key: meta_scalar(value) for key, value in parsed.meta.items()}


def stringify_blocks(blocks: list[dict[str, Any]]) -> str:
    return " ".join(
        stringify_inlines(block.get("c", []))
        for block in blocks
        if block.get("t") in {"Plain", "Para"}
    ).strip()


def stringify_inlines(inlines: Any) -> str:
    parts: list[str] = []

    def walk(node: Any) -> None:
        if isinstance(node, list):
            for item in node:
                walk(item)
            return
        if not isinstance(node, dict):
            return

        tag = node.get("t")
        value = node.get("c")

        if tag in {"Str", "Code", "Math"}:
            if isinstance(value, list):
                parts.append(str(value[-1]))
            else:
                parts.append(str(value))
        elif tag in {"Space", "SoftBreak", "LineBreak"}:
            parts.append(" ")
        elif tag in {"Emph", "Strong", "Strikeout", "SmallCaps", "Quoted", "Span", "Link"}:
            walk(value)
        elif tag == "Cite" and isinstance(value, list) and len(value) > 1:
            walk(value[1])
        else:
            walk(value)

    walk(inlines)
    return re.sub(r"\s+", " ", "".join(parts)).strip()


def slugify(text: str, *, fallback: str = "untitled") -> str:
    normalized = unicodedata.normalize("NFKD", text)
    ascii_text = normalized.encode("ascii", "ignore").decode("ascii")
    ascii_text = ascii_text.lower()
    ascii_text = re.sub(r"['’]", "", ascii_text)
    ascii_text = re.sub(r"[^a-z0-9]+", "-", ascii_text).strip("-")
    return ascii_text or fallback


def strip_known_prefix(stem: str) -> str:
    for prefix in (FINDING_PREFIX, TOPIC_PREFIX):
        if stem.startswith(prefix):
            return stem[len(prefix):]
    return stem


def first_words_from_blocks(blocks: list[dict[str, Any]], *, limit: int = 8) -> str:
    words: list[str] = []
    for block in blocks:
        if block.get("t") == "Header":
            words.extend(stringify_inlines(block["c"][2]).split())
        elif block.get("t") in {"Plain", "Para"}:
            words.extend(stringify_inlines(block.get("c", [])).split())
        if len(words) >= limit:
            break
    return " ".join(words[:limit])


def is_substantive(blocks: list[dict[str, Any]]) -> bool:
    return any(block.get("t") != "HorizontalRule" for block in blocks)


def demote_headings(blocks: list[dict[str, Any]], *, by: int = 1) -> list[dict[str, Any]]:
    demoted: list[dict[str, Any]] = []
    for block in blocks:
        if block.get("t") != "Header":
            demoted.append(block)
            continue
        level, attrs, inlines = block["c"]
        demoted.append({"t": "Header", "c": [level + by, attrs, inlines]})
    return demoted


def split_at_h1(blocks: list[dict[str, Any]], *, include_preamble: bool) -> list[Section]:
    sections: list[Section] = []
    current_title = ""
    current_blocks: list[dict[str, Any]] = []
    seen_h1 = False

    def flush() -> None:
        nonlocal current_title, current_blocks
        if not is_substantive(current_blocks):
            current_title = ""
            current_blocks = []
            return
        if current_title or include_preamble:
            title = current_title or "Preamble"
            sections.append(Section(len(sections) + 1, title, current_blocks))
        current_title = ""
        current_blocks = []

    for block in blocks:
        if block.get("t") == "Header" and block["c"][0] == 1:
            flush()
            seen_h1 = True
            current_title = stringify_inlines(block["c"][2]) or "Untitled"
            current_blocks = []
            continue
        current_blocks.append(block)

    flush()

    if not sections and is_substantive(blocks):
        # Fail useful: h1 is the contract, but a one-note fallback avoids empty runs
        # when testing older topic files.
        title = first_words_from_blocks(blocks) or "Untitled"
        sections.append(Section(1, title, blocks))

    return sections


def markdown_files(input_path: Path, *, recursive: bool) -> Iterable[Path]:
    if input_path.is_file():
        if input_path.suffix.lower() == ".md":
            yield input_path
        return

    pattern = "**/*.md" if recursive else "*.md"
    for path in sorted(input_path.glob(pattern)):
        if path.name.startswith("."):
            continue
        if any(part.startswith("_") for part in path.relative_to(input_path).parts[:-1]):
            continue
        if path.suffix.lower() == ".md":
            yield path


def write_default_template(path: Path, *, overwrite: bool = False) -> Path:
    target = path.expanduser().resolve()
    if target.exists() and not overwrite:
        fail(f"template already exists: {target}")
    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_text(DEFAULT_TEMPLATE, encoding="utf-8")
    return target


def ensure_template(template: str | Path | None) -> tuple[Path, tempfile.TemporaryDirectory[str] | None]:
    if template:
        path = Path(template).expanduser().resolve()
        if not path.exists():
            fail(f"template not found: {path}")
        return path, None

    td = tempfile.TemporaryDirectory()
    path = Path(td.name) / "fnd_template.pandoc.md"
    path.write_text(DEFAULT_TEMPLATE, encoding="utf-8")
    return path, td


def output_metadata(
    *,
    finding_id: str,
    topic_id: str,
    source_path: Path,
    source_heading: str,
    source_index: int,
    content_sha256: str,
    title: str,
    today: str,
    source_meta: dict[str, str],
) -> dict[str, Any]:
    # Keep the durable Obsidian record in frontmatter. Pull topic from source
    # frontmatter if present; otherwise infer from filename.
    topic = source_meta.get("topic") or topic_id
    return {
        "id": finding_id,
        "title": title,
        "topic": topic,
        "status": "imported",
        "verification_status": source_meta.get("verification_status") or "pending",
        "source_status": source_meta.get("source_status") or "unchecked",
        "source_topic_file": str(source_path),
        "source_heading": source_heading,
        "source_index": source_index,
        "input_sha256": content_sha256,
        "last_pipeline_run": "",
        "updated": today,
        "tags": ["finding", "imported_topic"],
    }


def index_markdown(rows: list[dict[str, str]]) -> str:
    lines = [
        "# Deconstructed Topic Findings",
        "",
        "| Topic | Finding | Source heading | Status |",
        "|---|---|---|---|",
    ]
    for row in rows:
        lines.append(
            f"| [[{row['topic']}]] | [[{row['finding']}]] | {row['heading']} | {row['status']} |"
        )
    lines.append("")
    return "\n".join(lines)


class TopicDeconstructor:
    """Fire CLI for deconstructing topic notes into finding notes."""

    def template(self, path: str = "fnd_template.pandoc.md", overwrite: bool = False) -> str:
        """Write the default Pandoc finding-note template."""
        written = write_default_template(Path(path), overwrite=overwrite)
        return f"wrote {written}"

    def deconstruct(
        self,
        input_path: str,
        output_path: str,
        recursive: bool = False,
        template: str | None = None,
        dry_run: bool = False,
        overwrite: bool = False,
        include_preamble: bool = False,
        write_index: bool = True,
        from_format: str = DEFAULT_FROM_FORMAT,
        to_format: str = DEFAULT_TO_FORMAT,
        wrap: str = "none",
    ) -> str:
        """Split topic files at H1 and write append-safe finding notes."""
        require_pandoc()

        src = Path(input_path).expanduser().resolve()
        out = Path(output_path).expanduser().resolve()
        if not src.exists():
            fail(f"input path does not exist: {src}")
        if wrap not in {"auto", "none", "preserve"}:
            fail("wrap must be one of: auto, none, preserve")

        template_path, template_tmp = ensure_template(template)
        today = dt.date.today().isoformat()
        rows: list[dict[str, str]] = []
        written_count = 0
        skipped_count = 0
        planned_count = 0

        if not dry_run:
            out.mkdir(parents=True, exist_ok=True)

        files = list(markdown_files(src, recursive=recursive))
        if not files:
            fail("no markdown files found")

        try:
            for source_path in files:
                parsed = parse_markdown(source_path, from_format=from_format)
                source_meta = source_frontmatter(parsed)

                topic_slug = slugify(strip_known_prefix(source_path.stem))
                inferred_topic_id = f"{TOPIC_PREFIX}{topic_slug}"
                topic_id = source_meta.get("id") or source_meta.get("topic") or inferred_topic_id

                sections = split_at_h1(parsed.blocks, include_preamble=include_preamble)
                print(f"{source_path.name}: {len(sections)} finding(s)")

                seen: dict[str, int] = {}
                for section in sections:
                    title_slug = slugify(section.title, fallback=f"finding-{section.index:03d}")
                    seen[title_slug] = seen.get(title_slug, 0) + 1
                    if seen[title_slug] > 1:
                        title_slug = f"{title_slug}-{seen[title_slug]}"

                    finding_slug = f"{topic_slug}__{section.index:03d}_{title_slug}"
                    filename = f"{FINDING_PREFIX}{finding_slug}.md"
                    out_file = out / filename
                    finding_id = out_file.stem

                    body_blocks = demote_headings(section.blocks, by=1)
                    body = ast_to_markdown(
                        blocks=body_blocks,
                        api_version=parsed.api_version,
                        to_format=to_format,
                        wrap=wrap,
                    )
                    content_hash = hashlib.sha256(body.encode("utf-8")).hexdigest()
                    metadata = output_metadata(
                        finding_id=finding_id,
                        topic_id=topic_id,
                        source_path=source_path,
                        source_heading=section.title,
                        source_index=section.index,
                        content_sha256=content_hash,
                        title=section.title,
                        today=today,
                        source_meta=source_meta,
                    )
                    note = render_template(
                        body_markdown=body,
                        metadata=metadata,
                        template_path=template_path,
                        from_format=from_format,
                        to_format=to_format,
                        wrap=wrap,
                    )

                    status = "planned" if dry_run else "written"
                    if dry_run:
                        print(f"  would write {out_file.name}")
                        planned_count += 1
                    elif out_file.exists() and not overwrite:
                        print(f"  skip existing {out_file.name}")
                        status = "skipped"
                        skipped_count += 1
                    else:
                        out_file.write_text(note, encoding="utf-8")
                        print(f"  wrote {out_file.name}")
                        written_count += 1

                    rows.append(
                        {
                            "topic": topic_id,
                            "finding": finding_id,
                            "heading": section.title.replace("|", "\\|") or "—",
                            "status": status,
                        }
                    )

            if write_index and not dry_run:
                index_path = out / "_deconstruction_index.md"
                index_path.write_text(index_markdown(rows), encoding="utf-8")
                print(f"wrote {index_path.name}")
        finally:
            if template_tmp is not None:
                template_tmp.cleanup()

        return (
            f"done: written={written_count}, skipped={skipped_count}, "
            f"planned={planned_count}, sources={len(files)}"
        )


if __name__ == "__main__":
    fire.Fire(TopicDeconstructor)

File: platform/extensions/pandoc/filters/import/deconstruct-topics.lua
-- deconstruct-topics.lua
--
-- Pandoc Lua filter for deconstructing topic files into finding notes.
--
-- Contract:
--   find Topics -name '*.md' -print0 |
--     xargs -0 -n1 pandoc -d deconstruct_topics -o Findings/_sink.md
--
-- The defaults file owns the Pandoc formats, data-dir, template name, and
-- deconstructor options. The output directory is the directory portion of the
-- outer Pandoc -o path unless deconstruct-output-dir is set in metadata.
--
-- Filesystem names are derived ONLY from the H1 heading text:
--   BCA IPO.md
--   BCA IPO 02.md
--
-- Pipeline identity is kept in frontmatter:
--   slug: fnd.<source-topic-slug>.<section-index>.<heading-slug>
--
-- The source topic filename is retained as metadata and as a safe tag. It is
-- not used as part of the output filename.

local path = pandoc.path

local stats = { written = 0, skipped = 0 }

local function fail(message)
  io.stderr:write("fatal: " .. message .. "\n")
  os.exit(1)
end

local function warn(message)
  io.stderr:write("warning: " .. message .. "\n")
end

local function meta_string(meta, key, fallback)
  local value = meta[key]
  if value == nil then return fallback or "" end
  return pandoc.utils.stringify(value)
end

local function required_meta(meta, key)
  local value = meta_string(meta, key, "")
  if value == "" then
    fail("missing required defaults metadata: " .. key)
  end
  return value
end

local function meta_bool(meta, key)
  local value = pandoc.text.lower(meta_string(meta, key, "false"))
  return value == "true" or value == "yes" or value == "1"
end

local function shell_quote(s)
  return "'" .. tostring(s):gsub("'", "'\\''") .. "'"
end

local function mkdir_p(dir)
  local ok = os.execute("mkdir -p " .. shell_quote(dir))
  if ok ~= true and ok ~= 0 then
    fail("could not create output directory: " .. dir)
  end
end

local function file_exists(file)
  local f = io.open(file, "r")
  if f then f:close(); return true end
  return false
end

local function write_text(file, text)
  local f = io.open(file, "w")
  if not f then fail("could not open output file: " .. file) end
  f:write(text)
  f:close()
end

local function strip_prefix(text, prefix)
  if prefix ~= "" and text:sub(1, #prefix) == prefix then
    return text:sub(#prefix + 1)
  end
  return text
end

local function slugify(text, fallback)
  local s = pandoc.text.lower(tostring(text or ""))
  s = s:gsub("[’']", "")
  s = s:gsub("[^a-z0-9]+", "-")
  s = s:gsub("^-+", ""):gsub("-+$", "")
  if s == "" then return fallback or "untitled" end
  return s
end

local function finding_slug(source_topic_slug, heading, index)
  local heading_slug = slugify(heading, string.format("finding-%03d", index))
  return string.format("fnd.%s.%03d.%s", source_topic_slug, index, heading_slug)
end

local function clean_filename_stem(text, fallback)
  local s = tostring(text or "")

  -- Keep the heading readable, but remove path separators and control chars.
  -- This deliberately does NOT slugify and does NOT add fnd/tpc prefixes.
  s = s:gsub("[\r\n\t]+", " ")
  s = s:gsub("[/\\]", "-")
  s = s:gsub(":", " -")
  s = s:gsub("[%c]", "")
  s = s:gsub("%s+", " ")
  s = s:gsub("^%s+", ""):gsub("%s+$", "")
  s = s:gsub("^%.", "_")

  if s == "" then return fallback or "Untitled" end
  return s
end

local function unique_output_file(out_dir, heading, index, used, overwrite)
  local base = clean_filename_stem(heading, string.format("Untitled %03d", index))
  local n = used[base] or 0

  while true do
    n = n + 1
    used[base] = n

    local output_stem = base
    if n > 1 then
      output_stem = string.format("%s %02d", base, n)
    end

    local output_file = path.normalize(path.join({ out_dir, output_stem .. ".md" }))

    if overwrite or not file_exists(output_file) then
      return output_file, output_stem
    end
  end
end

local function yaml_quote(value)
  local s = tostring(value or "")
  s = s:gsub("\\", "\\\\"):gsub('"', '\\"')
  return '"' .. s .. '"'
end

local function yaml_scalar(key, value)
  return key .. ": " .. yaml_quote(value) .. "\n"
end

local function yaml_number(key, value)
  return key .. ": " .. tostring(value) .. "\n"
end

local function yaml_tag(value)
  return "  - " .. yaml_quote(value) .. "\n"
end

local function today_iso()
  return os.date("%Y-%m-%d")
end

local function sha256(text)
  local ok, out = pcall(pandoc.pipe, "sha256sum", {}, text)
  if not ok then fail("sha256sum is not available on PATH") end
  return out:match("^([a-fA-F0-9]+)") or ""
end

local function demote_blocks(blocks, by)
  local out = {}
  for _, block in ipairs(blocks) do
    if block.t == "Header" then
      table.insert(out, pandoc.Header(block.level + by, block.content, block.attr))
    else
      table.insert(out, block)
    end
  end
  return out
end

local function is_substantive(blocks)
  for _, block in ipairs(blocks) do
    if block.t ~= "HorizontalRule" then return true end
  end
  return false
end

local function split_at_h1(blocks, include_preamble)
  local sections = {}
  local current_title = ""
  local current_blocks = {}

  local function flush()
    if not is_substantive(current_blocks) then
      current_title = ""
      current_blocks = {}
      return
    end

    if current_title ~= "" or include_preamble then
      table.insert(sections, {
        index = #sections + 1,
        title = current_title ~= "" and current_title or "Preamble",
        blocks = current_blocks,
      })
    end

    current_title = ""
    current_blocks = {}
  end

  for _, block in ipairs(blocks) do
    if block.t == "Header" and block.level == 1 then
      flush()
      current_title = pandoc.utils.stringify(block.content)
      if current_title == "" then current_title = "Untitled" end
      current_blocks = {}
    else
      table.insert(current_blocks, block)
    end
  end

  flush()

  if #sections == 0 and is_substantive(blocks) then
    fail("no level-1 headings found; H1 is the split contract")
  end

  return sections
end

local function first_input_file()
  if PANDOC_STATE and PANDOC_STATE.input_files and #PANDOC_STATE.input_files > 0 then
    return PANDOC_STATE.input_files[1]
  end
  return ""
end

local function output_dir(doc)
  local explicit = meta_string(doc.meta, "deconstruct-output-dir", "")
  if explicit ~= "" then return path.normalize(explicit) end

  if PANDOC_STATE and PANDOC_STATE.output_file and PANDOC_STATE.output_file ~= "" then
    return path.normalize(path.directory(PANDOC_STATE.output_file))
  end

  fail("no output filepath found; pass -o OUT/_sink.md or set deconstruct-output-dir")
end

local function user_data_dir()
  if PANDOC_STATE and PANDOC_STATE.user_data_dir and PANDOC_STATE.user_data_dir ~= "" then
    return PANDOC_STATE.user_data_dir
  end
  return ""
end

local function inner_pandoc_args(doc)
  local args = {}
  local data_dir = user_data_dir()
  if data_dir ~= "" then
    table.insert(args, "--data-dir")
    table.insert(args, data_dir)
  end

  table.insert(args, "--from")
  table.insert(args, required_meta(doc.meta, "deconstruct-from"))

  table.insert(args, "--to")
  table.insert(args, required_meta(doc.meta, "deconstruct-to"))

  table.insert(args, "--standalone")
  table.insert(args, "--template")
  table.insert(args, required_meta(doc.meta, "deconstruct-template"))

  table.insert(args, "--wrap")
  table.insert(args, meta_string(doc.meta, "deconstruct-wrap", "none"))

  return args
end

local function writer_format(doc)
  return required_meta(doc.meta, "deconstruct-to")
end

local function unescape_frontmatter(rendered)
  local head, fm, tail = rendered:match("^(%-%-%-\n)(.-)(\n%-%-%-\n.*)$")
  if not fm then return rendered end
  fm = fm:gsub("\\_", "_")
  return head .. fm .. tail
end

local function render_with_template(markdown, out_file, doc)
  local ok, rendered = pcall(pandoc.pipe, "pandoc", inner_pandoc_args(doc), markdown)
  if not ok then
    fail("inner pandoc failed for " .. out_file .. ": " .. tostring(rendered))
  end
  write_text(out_file, unescape_frontmatter(rendered))
end

local function section_input(metadata, body)
  local y = "---\n"
  y = y .. yaml_scalar("id", metadata.id)
  y = y .. yaml_scalar("slug", metadata.slug)
  y = y .. yaml_scalar("title", metadata.title)
  y = y .. yaml_scalar("kind", "finding")
  y = y .. yaml_scalar("topic", metadata.topic)
  y = y .. yaml_scalar("status", "imported")
  y = y .. yaml_scalar("verification_status", metadata.verification_status)
  y = y .. yaml_scalar("source_status", metadata.source_status)
  y = y .. yaml_scalar("source_topic_file", metadata.source_topic_file)
  y = y .. yaml_scalar("source_topic_title", metadata.source_topic_title)
  y = y .. yaml_scalar("source_topic_tag", metadata.source_topic_tag)
  y = y .. yaml_scalar("source_heading", metadata.source_heading)
  y = y .. yaml_number("source_index", metadata.source_index)
  y = y .. yaml_scalar("input_sha256", metadata.input_sha256)
  y = y .. yaml_scalar("last_pipeline_run", "")
  y = y .. yaml_scalar("updated", metadata.updated)
  y = y .. "tags:\n"
  y = y .. yaml_tag("finding")
  y = y .. yaml_tag("imported_topic")
  y = y .. yaml_tag(metadata.source_topic_tag)
  y = y .. "---\n\n"
  return y .. body .. "\n"
end

function Pandoc(doc)
  local source_path = meta_string(doc.meta, "deconstruct-source-path", first_input_file())
  if source_path == "" then fail("could not determine source path") end

  local out_dir = output_dir(doc)
  mkdir_p(out_dir)

  local finding_prefix = required_meta(doc.meta, "deconstruct-finding-prefix")
  local topic_prefix = required_meta(doc.meta, "deconstruct-topic-prefix")
  local overwrite = meta_bool(doc.meta, "deconstruct-overwrite")
  local include_preamble = meta_bool(doc.meta, "deconstruct-include-preamble")
  local sections = split_at_h1(doc.blocks, include_preamble)

  local source_stem = path.split_extension(path.filename(source_path))
  source_stem = strip_prefix(strip_prefix(source_stem, finding_prefix), topic_prefix)

  local source_topic_title = clean_filename_stem(source_stem, "Untitled Topic")
  local source_topic_slug = slugify(source_stem, "untitled-topic")
  local source_topic_tag = "topic/" .. source_topic_slug

  local inferred_topic = topic_prefix .. source_topic_slug
  local topic_id = meta_string(doc.meta, "id", "")
  if topic_id == "" then topic_id = meta_string(doc.meta, "topic", inferred_topic) end

  local used_output_stems = {}

  for _, section in ipairs(sections) do
    local out_file, output_stem = unique_output_file(
      out_dir,
      section.title,
      section.index,
      used_output_stems,
      overwrite
    )

    -- Human-readable filename stem doubles as note id. The machine slug below
    -- carries the fnd.* pipeline identity.
    local finding_id = output_stem

    local body_doc = pandoc.Pandoc(demote_blocks(section.blocks, 1), pandoc.Meta({}))
    local body = pandoc.write(body_doc, writer_format(doc), { wrap_text = "none" }):gsub("%s+$", "")
    local input_hash = sha256(body)

    local metadata = {
      id = finding_id,
      slug = finding_slug(source_topic_slug, section.title, section.index),
      title = section.title,
      topic = meta_string(doc.meta, "topic", topic_id),
      verification_status = meta_string(doc.meta, "verification_status", "pending"),
      source_status = meta_string(doc.meta, "source_status", "unchecked"),
      source_topic_file = source_path,
      source_topic_title = source_topic_title,
      source_topic_tag = source_topic_tag,
      source_heading = section.title,
      source_index = section.index,
      input_sha256 = input_hash,
      updated = today_iso(),
    }

    if file_exists(out_file) and not overwrite then
      warn("skip existing " .. out_file)
      stats.skipped = stats.skipped + 1
    else
      render_with_template(section_input(metadata, body), out_file, doc)
      io.stderr:write("wrote " .. out_file .. "\n")
      stats.written = stats.written + 1
    end
  end

  local summary = string.format(
    "deconstructed `%s`: written=%d skipped=%d sections=%d",
    source_path,
    stats.written,
    stats.skipped,
    #sections
  )

  return pandoc.Pandoc({ pandoc.Para({ pandoc.Str(summary) }) }, pandoc.Meta({}))
end
